internal/impl/kafka/franz_reader_ordered.go

```diff
@@ -0,0 +1,511 @@
+			delete(trackedTopic, lostPartition)
+		}
+		if len(trackedTopic) == 0 {
+			delete(c.topics, topicName)
+		}
+	}
+}
+
+func (c *partitionState) tallyActivePartitions(pausedPartitions map[string][]int32) (tally int) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	// This may not be 100% accurate, and perhaps even flakey, but as long as
+	// we're able to detect 0 active partitions then we're happy.
+	for topic, parts := range c.topics {
+		tally += (len(parts) - len(pausedPartitions[topic]))
+	}
+	return
+}
+
+//------------------------------------------------------------------------------
+
+// Connect to the kafka seed brokers.
+func (f *FranzReaderOrdered) Connect(ctx context.Context) error {
+	if f.partState != nil {
+		return nil
+	}
+
+	if f.shutSig.IsSoftStopSignalled() {
+		f.shutSig.TriggerHasStopped()
+		return service.ErrEndOfInput
+	}
+
+	var cl *kgo.Client
+	commitFn := func(r *kgo.Record) {}
+	if f.consumerGroup != "" {
+		commitFn = func(r *kgo.Record) {
+			if cl == nil {
+				return
+			}
+			cl.MarkCommitRecords(r)
+		}
+	}
+
+	checkpoints := newPartitionState(commitFn)
+
+	var clientOpts []kgo.Opt
+	clientOpts = append(clientOpts, f.clientOpts...)
+
+	if f.consumerGroup != "" {
+		clientOpts = append(clientOpts,
+			kgo.OnPartitionsRevoked(func(rctx context.Context, c *kgo.Client, m map[string][]int32) {
+				if commitErr := c.CommitMarkedOffsets(rctx); commitErr != nil {
+					f.log.Errorf("Commit error on partition revoke: %v", commitErr)
+				}
+				checkpoints.removeTopicPartitions(rctx, m)
+			}),
+			kgo.OnPartitionsLost(func(rctx context.Context, _ *kgo.Client, m map[string][]int32) {
+				// No point trying to commit our offsets, just clean up our topic map
+				checkpoints.removeTopicPartitions(rctx, m)
+			}),
+			kgo.OnPartitionsAssigned(func(rctx context.Context, _ *kgo.Client, m map[string][]int32) {
+				for topic, parts := range m {
+					for _, part := range parts {
+						// Adds the partition to our checkpointer
+						checkpoints.addRecords(rctx, topic, part, nil, f.cacheLimit)
+					}
+				}
+			}),
+			kgo.ConsumerGroup(f.consumerGroup),
+			kgo.AutoCommitMarks(),
+			kgo.AutoCommitInterval(f.commitPeriod),
+			kgo.WithLogger(&KGoLogger{f.log}),
+		)
+	}
+
+	var err error
+	if cl, err = kgo.NewClient(clientOpts...); err != nil {
+		return err
+	}
+
+	go func() {
+		defer func() {
+			cl.Close()
+			if f.shutSig.IsSoftStopSignalled() {
+				f.shutSig.TriggerHasStopped()
+			}
+		}()
+
+		closeCtx, done := f.shutSig.SoftStopCtx(context.Background())
+		defer done()
+
+		for {
+			// Using a stall prevention context here because I've realised we
+			// might end up disabling literally all the partitions and topics
+			// we're allocated.
+			//
+			// In this case we don't want to actually resume any of them yet so
+			// I add a forced timeout to deal with it.
+			stallCtx, pollDone := context.WithTimeout(closeCtx, time.Second)
+			fetches := cl.PollFetches(stallCtx)
```

Comment:
Not sure if this is an issue, but if we get a lot of backpressure, would it be possible to end up with a giant batch here? That was my reasoning when I used `PollRecords()` in the `redpanda_migrator` input, so I can pass in that `batch_size` parameter, which I probably should rename to `max_batch_size`.

Reply:
I'm not fully clued in to the mechanisms at play here, but I'm relying on the fetch max bytes, and fetch max partition bytes, fields for tuning the batch size. I'm definitely seeing huge batches in my test pipelines but benthos is actually handling it better than smaller batches.

Reply:
https://github.com/twmb/franz-go/blob/cea7aa5d803781e5f0162187795482ba1990c729/pkg/kgo/consumer.go#L401

`PollFetches` is just `PollRecords` but you get everything that is currently buffered. Relying on the configs for franz-go internal buffering is the way to go.

Reply:
Sounds good to me! Thank you!